File: algorithms/euler/problem31.clj
(defn update-for-coin [ways coin]
  (reduce
    (fn [w i] (assoc w i (+ (nth w i)
                            (nth w (- i coin)))))
    (vec ways)
    (range coin (count ways))))

(defn ways-to-make [amount]
  (let [ways-for-1p (map (fn [n] 1) (range (inc amount)))] 
    (loop [ways ways-for-1p coins [2 5 10 20 50 100 200]]
      (if (empty? coins)
        (last ways)
        (recur (update-for-coin ways (first coins)) (rest coins))))))

(time (println (ways-to-make 200)))

File: algorithms/euler/problem31.rb
COINS = [200, 100, 50, 20, 10, 5, 2, 1]

def ways_to_make(amount)
  ways = Array.new(amount + 1, 0)
  ways[0] = 1 # Kind of a hack to only count first coin once

  COINS.each do |coin|
    (amount/coin).times do |i|
      idx = (i+1)*coin      
      ways[idx] += ways[idx - coin]
    end    
  end

  return ways[amount]
end

puts ways_to_make(200)
